staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -381,6 +382,9 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 		}
 		if oldMeta.GetNamespace() == newMeta.GetNamespace() && oldMeta.GetName() == newMeta.GetName() {
 			if replaceExisting {
+				if reflect.DeepEqual(obj, existingObj) {
```

Comment:
strengthening the check here looks fine. but one trikcy/nitty part is that simply checking via reflection equality might not be completely mocking out apiserver's behavior:

1. checking via reflection will be failing when we check nil value against empty value. e.g. `metadata: {finalizers:[]}`  != `metadata:{}`. while for a running server, this is regarded as equal.

2. some api fields might be dropped or filled in running apiserver, which is to say, updating w/ a non-reflection-equal object might be regarded as equal object too.

wrt @caesarxuchao's previous comment, am not sure whether we should provide such a fine-grained mocking system in the first place

Reply:
This is meant to prevent update event when content does not change. It helps to detect bugs which may be hidden if we push will-not-happen update event in real apiserver.

For 2, modifications in apiserver (e.g. admission) are done and must be done in test to do testing. For example, fill default value or use fixtures with default value filled before updating to mock defaulting behavior in apiserver. It has no relation with watch mechanism IMO.

For 1, how kube-apiserver compares objects, `apiequality.Semantic.DeepEqual`? It is possible to use same way as kube-apiserver?

Reply:
> For 1, how kube-apiserver compares objects, apiequality.Semantic.DeepEqual? It is possible to use same way as kube-apiserver?

Watch is driven by actual updates in the serialized bytes stored in etcd, not by deepequal comparison like this, so the mechanism is not possible to reuse like this

Reply:
Thanks for pointing out! I'm not familiar with this. Looked at etcd3 Storage.Interface implementation, it uses runtime.Codec to encode object into bytes before comparing with original data in etcd.

I think we can mimic this logic. If bytes are equal, skip saving into store (which will do not trigger update event, same as etcd) and return original data.

https://github.com/kubernetes/kubernetes/blob/2601aadbb6cc7a4da0e91ceb74a021752ba7cf07/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L306-L329

Reply:
I think the current implementation with reflect.DeepEqual plus some documentation is sufficient. If a developer cares to write a test that depends on the fixture not sending the watch event, they should be capable enough to avoid the "nil != empty" case.

Serializers are tricky. I don't want to pull them in if possbile.

Reply:
I've added some comments here.

> If a developer cares to write a test that depends on the fixture not sending the watch event, they should be capable enough to avoid the "nil != empty" case.

Agree, and IMO at least this will not send update event on updating same object.